src/sqlfluff/cli/autocomplete.py

```diff
@@ -0,0 +1,51 @@
+"""CLI autocompletion methods."""
+import os
+import subprocess
+from typing import Dict, Optional
+
+
+def generate_autocomplete_script(
+    shell_type: str = "bash", save_path: Optional[str] = None
+) -> Dict[str, str]:
+    """Generate autocompletion script for specified shell type.
+
+    Args:
+        shell_type (str, optional): One of bash, zsh, or fish. Defaults to "bash".
+        save_path (Optional[str], optional): Location to save the autocompletion script.
+                                             Defaults to None.
+
+    Returns:
+        Dict[str, str]: Result dictionary object.
+    """
+    if save_path is None:
+        # Use default save_path if none is specified.
+        if shell_type.lower() == "fish":
+            save_path = "~/.config/fish/completions/.sqlfluff.fish"
+        else:
+            save_path = f"~/.config/sqlfluff/.sqlfluff-complete.{shell_type.lower()}"
+
+    if shell_type == "fish":
+        message = (
+            "Ensure the autocompletion script is saved (or symlinked) to "
+            "~/.config/fish/completions/.sqlfluff.fish, and then reload your shell."
+        )
+    else:
+        message = f"Please source {save_path} in your shell config file (e.g. ~.{shell_type}rc)."
+
+    # Ensure that parent config directory exists.
+    save_path = os.path.expanduser(save_path)
+    save_folder = os.path.dirname(save_path)
+    if not os.path.exists(save_folder):
+        os.makedirs(save_folder)
+
+    # Generate the autocompletion script.
+    subprocess.run(
+        [f"_SQLFLUFF_COMPLETE={shell_type.lower()}_source sqlfluff > {save_path}"],
```

Comment:
What is this doing exactly? What does it it run?

Reply:
It literally just runs SQLFluff with an environment variable `_SQLFLUFF_COMPLETE` that tells `click` to produce the relevant shell autocompletion script:
![image](https://user-images.githubusercontent.com/80432516/147847951-aec5ea5b-2c9a-47bc-a8cd-01cefe1e8e19.png)
the `> {save_path}` just saves the stdout to a file at the save path.

Docs: https://click.palletsprojects.com/en/8.0.x/shell-completion/?highlight=shell

Reply:
Would it be possible (desirable?) to do this using a Python function call rather than `subprocess.run()`? There's probably a tradeoff here, just thinking that the `subprocess.run()` reads a bit cryptic, and could a Python call be more readable. 🤔

Reply:
click doesn't provide an API to do it since it requires running via the CLI entry point to determine the relations between commands/sub-commands/args.

Relevant docs: https://click.palletsprojects.com/en/8.0.x/shell-completion/?highlight=shell#enabling-completion

Main reason for adding this as a sub-command to SQLFluff is to wrap the console command since I imagine the average non-technical SQLFluff user may be more comfortable that way.

That being said if we don't want the autocomplete sub-command we can just bin it and include a reference to the click shell completion section in the docs for those users who want to enable it. We should still keep the argument changes introduced in this PR as they are what tells autocompleter what to do for those who enable it.

Reply:
I'm good with adding it, was just curious about options for implementing it. Your explanation makes sense. Think it'd be worth summarizing any bits of the above discussion as code comments?

Reply:
I've reduced the scope of the PR to just fix CLI arguments/options so people can use the feature if they want. I'll make a later PR with a small doc note linking to click docs.